components/places/src/storage/maintenance.rs

```diff
@@ -0,0 +1,225 @@
+/// than this, some older visits will be deleted to free up space.  Pass in a 0 to skip this.
+pub fn run_maintenance_prune(conn: &PlacesDb, db_size_limit: u32) -> Result<RunMaintenanceMetrics> {
+    let db_size_before = conn.get_db_size()?;
+    let should_prune = db_size_limit > 0 && db_size_before > db_size_limit;
+    if should_prune {
+        let now = Timestamp::from(SystemTime::now());
+        history::prune_older_visits(conn)?;
+        prune_unknown_fields(conn, now)?;
+    }
+    let db_size_after = conn.get_db_size()?;
+    Ok(RunMaintenanceMetrics {
+        pruned_visits: should_prune,
+        db_size_before,
+        db_size_after,
+    })
+}
+
+/// Run maintenance on the places DB (vacuum step)
+///
+/// The `run_maintenance_*()` functions are intended to be run during idle time and will take steps
+/// to clean up / shrink the database.  They're split up so that we can time each one in the
+/// Kotlin wrapper code (This is needed because we only have access to the Glean API in Kotlin and
+/// it supports a stop-watch style API, not recording specific values).
+pub fn run_maintenance_vacuum(conn: &PlacesDb) -> Result<()> {
+    let auto_vacuum_setting: u32 = conn.query_one("PRAGMA auto_vacuum")?;
+    if auto_vacuum_setting == 2 {
+        // Ideally, we run an incremental vacuum to delete 2 pages
+        conn.execute_one("PRAGMA incremental_vacuum(2)")?;
+    } else {
+        // If auto_vacuum=incremental isn't set, configure it and run a full vacuum.
+        log::warn!(
+            "run_maintenance_vacuum: Need to run a full vacuum to set auto_vacuum=incremental"
+        );
+        conn.execute_one("PRAGMA auto_vacuum=incremental")?;
+        conn.execute_one("VACUUM")?;
+    }
+    Ok(())
+}
+
+/// Run maintenance on the places DB (optimize step)
+///
+/// The `run_maintenance_*()` functions are intended to be run during idle time and will take steps
+/// to clean up / shrink the database.  They're split up so that we can time each one in the
+/// Kotlin wrapper code (This is needed because we only have access to the Glean API in Kotlin and
+/// it supports a stop-watch style API, not recording specific values).
+pub fn run_maintenance_optimize(conn: &PlacesDb) -> Result<()> {
+    conn.execute_one("PRAGMA optimize")?;
+    Ok(())
+}
+
+/// Run maintenance on the places DB (checkpoint step)
+///
+/// The `run_maintenance_*()` functions are intended to be run during idle time and will take steps
+/// to clean up / shrink the database.  They're split up so that we can time each one in the
+/// Kotlin wrapper code (This is needed because we only have access to the Glean API in Kotlin and
+/// it supports a stop-watch style API, not recording specific values).
+pub fn run_maintenance_checkpoint(conn: &PlacesDb) -> Result<()> {
+    conn.execute_one("PRAGMA wal_checkpoint(PASSIVE)")?;
+    Ok(())
+}
+
+fn prune_unknown_fields(conn: &PlacesDb, now: Timestamp) -> Result<()> {
```

Comment:
This is a nice solution! Would be nice to have some comments of why this exists and why we need to prune (I'm assuming it's related to the immutable visits discussion). 

That being said, I'm not entirely sure how I feel of separating the maintenance of pruning unknown_fields -- wouldn't they get removed with the visits their associated with (in `prune_older_visits`) or is this used for something else? Comments here might also lessen the confusion of this.

Reply:
Sorry, I meant to add a comment here and totally forgot about it.  But now that I'm writing it down, I'm having second thoughts.  Maybe we shouldn't special case the `unknown_fields` and just continue to use the existing pruning code.

The reasoning behind the `prune_unknown_fields` is that we only sync the most recent 20 visits for each URL, so storing `unknown_fields` for visits older than that seems like a waste of space.  However, that's not really true:
  - As you point out, the client will upgrade and at that point it will want to migrate the `unknown_fields` data and start using it.
  - If you delete recent visits, then that will push older URLs into the most recent 20.  (This is related to https://github.com/mozilla/application-services/issues/5472).

Part of the motivation here is there's a lot of visits in the DB and `unknown_fields` is space inefficient.  You want to add an enum column that could be stored in as an integer, you end up storing an entire JSON string.  Maybe we can address this another way though:
  - Lets say that we are on nightly version v120 and want to add a new feature that requires a new field
  - Before starting work on that feature, we add a migration to store the new field in a column and round-trip the data.
  - The migration should be pretty trivial, so we uplift it into v119 and an ESR patch ahead of the v120 release.
  - That way users who are doing a reasonable job keeping up-to-date don't have to deal with the space penalty.